Fix: issues in Onyx bump due to removal of null values in cache

chrispader opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

@mountiny 

### Details
<!-- Explanation of the change or anything fishy that is going on -->

This PR fixes minor issues in Onyx arising from [this Onyx bump PR](https://github.com/margelo/expensify-app-fork/pull/20).

- Never set `undefined` values in cache and storage, when `undefined` is passed to `Onyx.set`/`Onyx.merge`/...
-  Remove duplicate "key removal" logic in `Onyx.multiSet`
- `OnyxUtils.getCachedCollection` should only return `undefined` keys that are actually set in cache. (We need strict objects with _not set_ keys rather then keys that are `undefined` when not existing)
- Fix issues with `deepEqual` comparisons in `keyChanged` and `keysChanged` (and prefer easier-to-read early-returns) (improve naming of some variables and make them more consistent)
- Due to the TS changes, i noticed that in `mergeObject` we're trying to access a key on `target`, even though `target` might not even be an object
- Fix an issue in `withOnyx` where we check if all state entries in `tempState` are non-`undefined`, rather than checking if the keys exist. This caused components to never exit the loading state in `withOnyx` and therefore never render

The PR also adds some TS type changes and improvements related to Onyx input values, that also stem from this Onyx bump PR.

### Related Issues
<!-- Please replace GH_LINK with the link to the GitHub issue this Pull Request is related to -->
https://github.com/margelo/expensify-app-fork/pull/20

### Automated Tests
<!---
Most changes to Onyx should have accompanying tests. Describe the tests you added or if no tests were added an explanation about why one was not needed.
--->

### Manual Tests
<!---
Each set of changes should be tested against the Expensify/App repo on all platforms to verify it does not break our staging App.
--->

### Author Checklist

- [X] I linked the correct issue in the `### Related Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

chrispader commented:
@mountiny updated the PR description. This PR should fix all remaining issues in the Onyx bump PR (so far).

@blazejkustra @fabioh8010 the TS changes are mostly related to similar changes i made in a previous PR, related to Onyx input types (`Onyx.merge`, `Onyx.set`)

chrispader commented:
ready for review! 🚀

chrispader commented:
@mountiny working on adding new tests for these changes today!

fabioh8010 commented on lib/types.ts:
````diff
@@ -328,25 +312,73 @@ type Mapping<TKey extends OnyxKey> = ConnectOptions<TKey> & {
+/**
+ * Represents a single Onyx input value, that can be either `TOnyxValue` or `null` if the key should be deleted.
+ *
+ * It can be used to specify data retrieved from Onyx e.g. `withOnyx` HOC mappings.
+ *
+ * @example
+ * ```ts
+ * import Onyx, {OnyxEntry, withOnyx} from 'react-native-onyx';
+ *
+ * type OnyxProps = {
+ *     userAccount: OnyxEntry<Account>;
+ * };
+ *
+ * type Props = OnyxProps & {
+ *     prop1: string;
+ * };
+ *
+ * function Component({prop1, userAccount}: Props) {
+ *     // ...
+ * }
+ *
+ * export default withOnyx<Props, OnyxProps>({
+ *     userAccount: {
+ *         key: ONYXKEYS.ACCOUNT,
+ *     },
+ * })(Component);
+ * ```
````
Is this example right? I'm not seeing `OnyxInputValue` being used there.

fabioh8010 commented on lib/types.ts:
```diff
@@ -328,25 +312,73 @@ type Mapping<TKey extends OnyxKey> = ConnectOptions<TKey> & {
+ * Setting a key to `null` will remove the key from the store.
+ * `undefined` is not allowed for setting values, because it will have no effect on the data.
+ */
+// type OnyxInput<TOnyxValue> = TOnyxValue | null;
```
```suggestion
```

fabioh8010 reviewed (commented):
@chrispader Nice work, left few comments!

Could you apply these changes too? We don't need them anymore since we get rid of `null`.

```diff
diff --git a/lib/useOnyx.ts b/lib/useOnyx.ts
index 958170b..230cbaf 100644
--- a/lib/useOnyx.ts
+++ b/lib/useOnyx.ts
@@ -2,21 +2,11 @@ import {deepEqual} from 'fast-equals';
 import {useCallback, useEffect, useRef, useSyncExternalStore} from 'react';
 import type {IsEqual} from 'type-fest';
 import OnyxUtils from './OnyxUtils';
-import type {CollectionKeyBase, KeyValueMapping, NonNull, OnyxCollection, OnyxEntry, OnyxKey, Selector} from './types';
+import type {CollectionKeyBase, OnyxCollection, OnyxKey, OnyxValue, Selector} from './types';
 import useLiveRef from './useLiveRef';
 import usePrevious from './usePrevious';
 import Onyx from './Onyx';
 
-/**
- * Represents a Onyx value that can be either a single entry or a collection of entries, depending on the `TKey` provided.
- * It's a variation of `OnyxValue` type that is read-only and excludes the `null` type.
- */
-type UseOnyxValue<TKey extends OnyxKey> = string extends TKey
-    ? unknown
-    : TKey extends CollectionKeyBase
-    ? NonNull<OnyxCollection<KeyValueMapping[TKey]>>
-    : NonNull<OnyxEntry<KeyValueMapping[TKey]>>;
-
 type BaseUseOnyxOptions = {
     /**
      * Determines if this key in this subscription is safe to be evicted.
@@ -55,7 +45,7 @@ type UseOnyxOptions<TKey extends OnyxKey, TReturnValue> = BaseUseOnyxOptions & U
 
 type FetchStatus = 'loading' | 'loaded';
 
-type CachedValue<TKey extends OnyxKey, TValue> = IsEqual<TValue, UseOnyxValue<TKey>> extends true ? TValue : TKey extends CollectionKeyBase ? NonNullable<OnyxCollection<TValue>> : TValue;
+type CachedValue<TKey extends OnyxKey, TValue> = IsEqual<TValue, OnyxValue<TKey>> extends true ? TValue : TKey extends CollectionKeyBase ? NonNullable<OnyxCollection<TValue>> : TValue;
 
 type ResultMetadata = {
     status: FetchStatus;
@@ -67,15 +57,15 @@ function getCachedValue<TKey extends OnyxKey, TValue>(key: TKey, selector?: Sele
     return OnyxUtils.tryGetCachedValue(key, {selector}) as CachedValue<TKey, TValue> | undefined;
 }
 
-function useOnyx<TKey extends OnyxKey, TReturnValue = UseOnyxValue<TKey>>(
+function useOnyx<TKey extends OnyxKey, TReturnValue = OnyxValue<TKey>>(
     key: TKey,
     options?: BaseUseOnyxOptions & UseOnyxInitialValueOption<TReturnValue> & Required<UseOnyxSelectorOption<TKey, TReturnValue>>,
 ): UseOnyxResult<TKey, TReturnValue>;
-function useOnyx<TKey extends OnyxKey, TReturnValue = UseOnyxValue<TKey>>(
+function useOnyx<TKey extends OnyxKey, TReturnValue = OnyxValue<TKey>>(
     key: TKey,
     options?: BaseUseOnyxOptions & UseOnyxInitialValueOption<NoInfer<TReturnValue>>,
 ): UseOnyxResult<TKey, TReturnValue>;
-function useOnyx<TKey extends OnyxKey, TReturnValue = UseOnyxValue<TKey>>(key: TKey, options?: UseOnyxOptions<TKey, TReturnValue>): UseOnyxResult<TKey, TReturnValue> {
+function useOnyx<TKey extends OnyxKey, TReturnValue = OnyxValue<TKey>>(key: TKey, options?: UseOnyxOptions<TKey, TReturnValue>): UseOnyxResult<TKey, TReturnValue> {
     const connectionIDRef = useRef<number | null>(null);
     const previousKey = usePrevious(key);
 

```

blazejkustra reviewed (commented):
Type changes look good to me! I really like the idea to replace `null` with `undefined` 👍 

Do I understand correctly that `null` will still be used but only when we want to remove a property/whole onyx value?

chrispader commented:
> Do I understand correctly that `null` will still be used but only when we want to remove a property from an onyx value/remove whole onyx value?

@blazejkustra yes exactly. `null` now means "delete that key" (or sub-property of some key.

Onyx will never persist and return `null` values though, but always return `undefined` from `Onyx.connect` callback, `withOnyx` and `useOnyx`, if a key doesn't exist in cache/storage

chrispader commented:
@fabioh8010 @blazejkustra just added one more type that we're gonna use in E/App, so you can probably ignore reviewing the last two commit, because it's just a TS type export

fabioh8010 commented on lib/types.ts:
```diff
@@ -318,6 +318,11 @@ type Mapping<TKey extends OnyxKey> = ConnectOptions<TKey> & {
+/**
+ * Represents an Onyx collection input, that can be either a record of `TOnyxValue`s or `null` if the key should be deleted.
+ */
+type OnyxCollectionInput<TOnyxValue> = OnyxInputValue<Record<string, TOnyxValue | null>>;
```
```suggestion
type OnyxCollectionInputValue<TOnyxValue> = OnyxInputValue<Record<string, TOnyxValue | null>>;
```

Just to be aligned with `OnyxInputValue` pattern

fabioh8010 commented:
@chrispader I was checking the last changes and I found an TS issue with `mergeCollection`. Now it's being possible to pass a different collection member key related to the one specified in the function:

```ts
    Onyx.mergeCollection(ONYXKEYS.COLLECTION.DOWNLOAD, {
        [`${ONYXKEYS.COLLECTION.DOWNLOAD}${'attachment1'}` as const]: {isDownloading: true},
        [`${ONYXKEYS.COLLECTION.DOWNLOAD}${'attachment3'}` as const]: {isDownloading: true},
        [`${ONYXKEYS.COLLECTION.REPORT}${'report1'}` as const]: {reportID: 'account'}, // TS is allowing to pass a report item with a download object, previously it would raise an error
    });
```

I did some changes and here's my suggested fix, could you check if they make sense to your changes?

```diff
diff --git a/lib/Onyx.ts b/lib/Onyx.ts
index 74ff100..d300690 100644
--- a/lib/Onyx.ts
+++ b/lib/Onyx.ts
@@ -16,13 +16,14 @@ import type {
     OnyxInputKeyValueMapping,
     OnyxCollection,
     OnyxKey,
-    OnyxMergeCollectionInput,
     OnyxMergeInput,
     OnyxMultiSetInput,
     OnyxSetInput,
     OnyxUpdate,
     OnyxValue,
     OnyxInput,
+    MergeCollection,
+    NullishDeep,
 } from './types';
 import OnyxUtils from './OnyxUtils';
 import logMessages from './logMessages';
@@ -423,12 +424,12 @@ function merge<TKey extends OnyxKey>(key: TKey, changes: OnyxMergeInput<TKey>):
  * @param collectionKey e.g. `ONYXKEYS.COLLECTION.REPORT`
  * @param collection Object collection keyed by individual collection member keys and values
  */
-function mergeCollection<TKey extends CollectionKeyBase>(collectionKey: TKey, collection: OnyxMergeCollectionInput<TKey>): Promise<void> {
+function mergeCollection<TKey extends CollectionKeyBase, TMap>(collectionKey: TKey, collection: MergeCollection<TKey, TMap, NullishDeep<KeyValueMapping[TKey]>>): Promise<void> {
     if (typeof collection !== 'object' || Array.isArray(collection) || utils.isEmptyObject(collection)) {
         Logger.logInfo('mergeCollection() called with invalid or empty value. Skipping this update.');
         return Promise.resolve();
     }
-    const mergedCollection = collection;
+    const mergedCollection: OnyxInputKeyValueMapping = collection;
 
     // Confirm all the collection keys belong to the same parent
     let hasCollectionKeyCheckFailed = false;
diff --git a/lib/types.ts b/lib/types.ts
index 3b098b0..8dea30c 100644
--- a/lib/types.ts
+++ b/lib/types.ts
@@ -252,7 +252,7 @@ type NullishObjectDeep<ObjectType extends object> = {
  * Also, the `TMap` type is inferred automatically in `mergeCollection()` method and represents
  * the object of collection keys/values specified in the second parameter of the method.
  */
-type Collection<TKey extends CollectionKeyBase, TValue, TMap = never> = {
+type MergeCollection<TKey extends CollectionKeyBase, TMap, TValue> = {
     [MapK in keyof TMap]: MapK extends `${TKey}${string}`
         ? MapK extends `${TKey}`
             ? never // forbids empty id
@@ -434,7 +434,7 @@ type GenericFunction = (...args: any[]) => any;
 
 export type {
     BaseConnectOptions,
-    Collection,
+    MergeCollection,
     CollectionConnectCallback,
     CollectionConnectOptions,
     CollectionKey,
diff --git a/tests/unit/useOnyxTest.ts b/tests/unit/useOnyxTest.ts
index d45eef9..13acc5e 100644
--- a/tests/unit/useOnyxTest.ts
+++ b/tests/unit/useOnyxTest.ts
@@ -3,6 +3,7 @@ import type {OnyxEntry} from '../../lib';
 import Onyx, {useOnyx} from '../../lib';
 import waitForPromisesToResolve from '../utils/waitForPromisesToResolve';
 import StorageMock from '../../lib/storage';
+import type GenericCollection from '../utils/GenericCollection';
 
 const ONYXKEYS = {
     TEST_KEY: 'test',
@@ -167,7 +168,7 @@ describe('useOnyx', () => {
                 [`${ONYXKEYS.COLLECTION.TEST_KEY}entry1`]: {id: 'entry1_id', name: 'entry1_name'},
                 [`${ONYXKEYS.COLLECTION.TEST_KEY}entry2`]: {id: 'entry2_id', name: 'entry2_name'},
                 [`${ONYXKEYS.COLLECTION.TEST_KEY}entry3`]: {id: 'entry3_id', name: 'entry3_name'},
-            });
+            } as GenericCollection);
 
             const {result} = renderHook(() =>
                 useOnyx(ONYXKEYS.COLLECTION.TEST_KEY, {
@@ -199,7 +200,7 @@ describe('useOnyx', () => {
                 [`${ONYXKEYS.COLLECTION.TEST_KEY}entry1`]: {id: 'entry1_id', name: 'entry1_name'},
                 [`${ONYXKEYS.COLLECTION.TEST_KEY}entry2`]: {id: 'entry2_id', name: 'entry2_name'},
                 [`${ONYXKEYS.COLLECTION.TEST_KEY}entry3`]: {id: 'entry3_id', name: 'entry3_name'},
-            });
+            } as GenericCollection);
 
             const {result} = renderHook(() =>
                 useOnyx(ONYXKEYS.COLLECTION.TEST_KEY, {
diff --git a/tests/utils/GenericCollection.ts b/tests/utils/GenericCollection.ts
index b00fce6..170f17e 100644
--- a/tests/utils/GenericCollection.ts
+++ b/tests/utils/GenericCollection.ts
@@ -1,5 +1,5 @@
-import type {Collection} from '../../lib/types';
+import type {MergeCollection} from '../../lib/types';
 
-type GenericCollection = Collection<string, unknown, unknown>;
+type GenericCollection = MergeCollection<string, unknown, unknown>;
 
 export default GenericCollection;

```

chrispader commented:
> @chrispader I was checking the last changes and I found an TS issue with `mergeCollection`. Now it's being possible to pass a different collection member key related to the one specified in the function:
> 
> ```ts
>     Onyx.mergeCollection(ONYXKEYS.COLLECTION.DOWNLOAD, {
>         [`${ONYXKEYS.COLLECTION.DOWNLOAD}${'attachment1'}` as const]: {isDownloading: true},
>         [`${ONYXKEYS.COLLECTION.DOWNLOAD}${'attachment3'}` as const]: {isDownloading: true},
>         [`${ONYXKEYS.COLLECTION.REPORT}${'report1'}` as const]: {isDownloading: true}, // TS is allowing to pass a report item with a download object, previously it would raise an error
>     });
> ```
> 
> **EDIT:** It seems that the issue is fixed if I revert [these changes](https://github.com/Expensify/react-native-onyx/pull/558/commits/4e893aa277ceacce0a903afb765ed1e724ec0d5a), why do we need them @chrispader ?

@fabioh8010 yes you're right. I thought we could pass the type of the collection to `mergeCollection` as a generic parameter, though the collection types are actually defined in E/App in a  module declaration`.d.ts` and not in Onyx itself.

Reverted the changes and added back the `@ts-expect-error` flags in `useOnyxTest`

fabioh8010 reviewed (approved):
🚀

mountiny reviewed (approved):
Thank you for the changes, all bugs found in the App PR has been addressed it seems so lets get this out so we can make a new build for Applause
